### DISCUSSION TITLE
LPC55S16JBD64 CAN is not working after configuring board files

### DISCUSSION THREAD

**Discussion by aravinth-kalai** [zephyrproject-rtos/zephyr#62566] [Q&A] [answered]
Hello Everyone , 

I have already tested CAN feature with LPC55S16JBD100 (100 pin) now,  I'm trying to use CAN in LPC55S16JBD64  controller, so I configured board files and if I try to send or receive something the data is not showing on CAN bus , I'm using TJA1462AT driver [schematic refrence](https://global.discourse-cdn.com/nvidia/original/3X/0/f/0f4600b30b661a75e1b23096c5e00cf66bbde4e3.png) for CAN, I'll share my board file configurations and sample code for your reference.

**.dtsi file** 
````#include <nxp/lpc/LPC55S16JBD64-pinctrl.h>

/ {
	chosen {
		zephyr,sram = &sram0;
		zephyr,flash = &flash0;
		zephyr,code-partition = &sramx;
		zephyr,console = &flexcomm0;
		zephyr,shell-uart = &flexcomm0;
		zephyr,canbus = &can0;
	};
aliases{
		cancom = &can0;
	};

&can0 {
	status = "okay";
	bus-speed = <125000>;
	bus-speed-data = <125000>;
	pinctrl-0 = <&pinmux_can0>;
	pinctrl-names = "default";

	can-transceiver {
		max-bitrate = <1000000>;
	};
};
&pinctrl {
	pinmux_can0: pinmux_can0 {
		group0 {
			pinmux = <CAN0_RD_PIO1_3>,
				<CAN0_TD_PIO1_2>;
			slew-rate = "standard";
		};
	};

```` 

**Kconfig.board**
````# LPCXpresso55S16 board

# Copyright (c) 2020 Henrik Brix Andersen <henrik@brixandersen.dk>
# SPDX-License-Identifier: Apache-2.0

config BOARD_LPC55s16_EVAL 
	bool "LPC55s16 Eval board" 
	depends on SOC_SERIES_LPC55XXX
	select SOC_PART_NUMBER_LPC55S16JBD64
````
and I have added few lines in the **Kconfig.Soc**

soc\arm\nxp_lpc\lpc55xxx\Kconfig.soc start with line number 83, I have added the few lines to avoid build errors

````
if SOC_SERIES_LPC55XXX

config SOC_PART_NUMBER_LPC55S06JBD64
	bool

config SOC_PART_NUMBER_LPC55S16JBD64  //added this line
	bool                   //added this line

config SOC_PART_NUMBER_LPC55S16JBD100
	bool

config SOC_PART_NUMBER_LPC55S28JBD100
	bool

config SOC_PART_NUMBER_LPC55S36JBD100
	bool

config SOC_PART_NUMBER_LPC55S69JBD100
	bool

config SOC_PART_NUMBER_LPC55S69JET98
	bool

config SOC_PART_NUMBER_LPC55XXX
	string
	default "LPC55S06JBD64"  if SOC_PART_NUMBER_LPC55S06JBD64
	default "LPC55S16JBD64"  if SOC_PART_NUMBER_LPC55S16JBD64 // added this line
	default "LPC55S16JBD100" if SOC_PART_NUMBER_LPC55S16JBD100
	default "LPC55S28JBD100" if SOC_PART_NUMBER_LPC55S28JBD100
	default "LPC55S36JBD100" if SOC_PART_NUMBER_LPC55S36JBD100
	default "LPC55S69JBD100" if SOC_PART_NUMBER_LPC55S69JBD100
	default "LPC55S69JET98"  if SOC_PART_NUMBER_LPC55S69JET98

	help
	  This string holds the full part number of the SoC. It is a hidden
	  option that you should not set directly. The part number selection
	  choice defines the default value for this string.
````

Sample code to test micro controller receive functionality 
````
#include <drivers/can.h>
#include <zephyr/devicetree.h>
#include  <zephyr/kernel.h>
CAN_MSGQ_DEFINE(my_can_msgq, 8);

void main()
{
   int result = 2;
   const struct device *dev = DEVICE_DT_GET(DT_ALIAS(cancom));
   int err;
    if (!device_is_ready(dev)) {
		printk("CAN device not ready\n");
		return;
	}
    else
    {
        printk("CAN device ready\n");
    }

    err = can_start(dev);
	if (err != 0) {
		printk("Error starting CAN controller (err %d)\n", err);
		return;
	}
    else
    {
       printk("CAN device start okay\n");
    }
while(1)
{
    result = k_msgq_get(&my_can_msgq, &rx_frame, K_FOREVER);
    if(result == 0)
    {
        printk("Message Received\n");
    }
    else
    {
        printk("failed to receive can frame\n");
    }
}
}
````
If I provide input from PEAK CAN with 125kbps baud rate

I'm getting the following prints 
CAN device ready
CAN device start okay

but I'm not getting any prints related to Message Received  or failed to receive can frame, am I missing something , please help me to fix this issue.

**Note : Also I have probed with oscilloscope with CAN pins (CAN0_RD_PIO1_3 and CAN0_TD_PIO1_2), I'm not getting any signals from those pins, but during the reception I can see data at the RX pin of transceiver IC (sent from PEAK CAN)**

**Comment by DerekSnell** [comment, marked as answer]
Hi @aravinth-kalai ,
The difference between LPC55S16JBD100 and LPC55S16JBD64 is just the package.  It is the same SoC, just less pins offered in the 64-pin package.  So the SoC configuration for the Zephyr firmware should be basically the same, except maybe the pins used in the pinctrl.

You confirmed the CAN example succeeds on the LPC55S16JBD100 device, and I assume that is using the NXP board lpcxpresso55s16.  Your board with the 64-pin device must be a custom board.  Have you already ruled out any hardware issues with the custom board?  Or hardware differences that require a different config in Zephyr?

Since you have this working with the 100-pin package, I would build a single firmware image to run on both devices.  And use that to test the other board.  If it runs fine on the 100-pin device, but fails on the other board, it points to a hardware issue or some hardware difference of that board that needs to be configured differently.  You should be able to build this image using the LPC55S16JBD100 device name, since that hardware is working for you, but pick pins that are in common with the 64-pin package.

Another potential option to consider is to use a [CAN example](https://github.com/nxp-mcuxpresso/mcux-sdk-examples/tree/main/lpcxpresso55s16/driver_examples/mcan) from the MCUXpresso SDK to confirm your hardware.

Let us know what you find

**Reply by aravinth-kalai** [reply]
Hi @DerekSnell , 

Thanks for your reply, 

Sorry for late reply, 

I tried with MCUXpresso SDK mcan_interrupt_transfer example by changing pin numbers in pin_mux.c file
```
    const uint32_t port1_pin3_config = (/* Pin is configured as CAN0_RD */
                                         IOCON_PIO_FUNC9 |
                                         /* Selects pull-up function */
                                         IOCON_PIO_MODE_PULLUP |
                                         /* Standard mode, output slew rate control is enabled */
                                         IOCON_PIO_SLEW_STANDARD |
                                         /* Input function is not inverted */
                                         IOCON_PIO_INV_DI |
                                         /* Enables digital function */
                                         IOCON_PIO_DIGITAL_EN |
                                         /* Open drain is disabled */
                                         IOCON_PIO_OPENDRAIN_DI);
    /* PORT1 PIN3 (coords: 41) is configured as CAN0_RD */
    IOCON_PinMuxSet(IOCON, 1U, 3U, port1_pin3_config);

    const uint32_t port1_pin2_config = (/* Pin is configured as CAN0_TD */
                                         IOCON_PIO_FUNC9 |
                                         /* Selects pull-up function */
                                         IOCON_PIO_MODE_PULLUP |
                                         /* Standard mode, output slew rate control is enabled */
                                         IOCON_PIO_SLEW_STANDARD |
                                         /* Input function is not inverted */
                                         IOCON_PIO_INV_DI |
                                         /* Enables digital function */
                                         IOCON_PIO_DIGITAL_EN |
                                         /* Open drain is disabled */
                                         IOCON_PIO_OPENDRAIN_DI);
    /* PORT1 PIN2 (coords: 85) is configured as CAN0_TD */
    IOCON_PinMuxSet(IOCON, 1U, 2U, port1_pin2_config);
```
After flashing I'm getting these prints 
```
Please select local node as A or B:
Note: Node B should start first.
Node:B
B
Start to Wait data from Node A
```
then I set baud rate as 500kbps in PCAN-view tool and I try to send data with frame ID of 0X321 , the tool shows bus heavy (if try with zephyr also getting same bus heavy error), also I'm not getting any print in console.

Note: Same code if I flash it on lpcxpresso55s16 EVAL board by changing pin number PIO1_22 as CAN0_RD and PIO1_27 as CAN0_TD then if I try to send some data using PCAN-view with the same baudrate (500kbps), I can see the same data echo in PCAN_view tool but the same code is not working with LPC55S16JBD64 custom board 

Also we verified the hardware , there is no issue with hardware , Please help me to fix the issue.

**Comment by DerekSnell** [comment]
Hi @aravinth-kalai ,
These GitHub discussions are specific to Zephyr.  And based on your details, you are seeing a similar CAN issue even when not using Zephyr.  So it seems this issue is related to a hardware difference on your custom board or with the different package.

I recommend you post you question at NXP's support forum https://community.nxp.com, in your case specifically at [LPC Microcontrollers](https://community.nxp.com/t5/LPC-Microcontrollers/bd-p/lpc).  Then your issue can be assigned to NXP support engineers focused on this CAN peripheral, and this MCU.  And with them, I would focus on your MCUXpresso SDK project, to simplify the firmware differences by removing Zephyr.  Once you find the root cause and resolve it, you can update your Zephyr board files to match.  

Once you create your post at https://community.nxp.com, please add that link back to this Discussion.  Then the rest of us with interest in this can follow the solution.

Best regards
